### PR TITLE
[depends: #708] Write Stack for storing pointer values

### PR DESCRIPTION
**Pull request by thk123**
A more sophisticated mechanism for storing pointers. It supports pointers to things like:

```C
&a[1]
&s.comp
p + 1 // p = &a[4]
```

It does this by taking an expression and converting it into a stack that can then be converted back to an expression. 

This is dependent on the variable sensitivity domain #708 
It also uses the improvements to the unit test framework #1111 :white_check_mark: 

Real diff: https://github.com/diffblue/cbmc/pull/1120/files/964b8c841bebe266643159eb377eadef89df4336..0fcf28001b576ff4ed9aebf06dc678ce3c1a61e8

### CODE REVIEW

**Comment by tautschnig** [comment]
Is there published work that describes this approach? What I am wondering about is how branching (any non-linear control-flow) interacts with those stacks?

File: unit/analyses/variable-sensitivity/write_stack.cpp [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+          auto stack=write_stackt(in_expr, environment, ns);
+          REQUIRE_FALSE(stack.is_top_value());
+          const exprt &out_expr=stack.to_expression();
+
+          CAPTURE(expr2c(out_expr, ns));
+
+          REQUIRE(out_expr.id()==ID_address_of);
+          const member_exprt &member_exrp=
+            require_exprt::require_member(out_expr.op0(), "comp");
+          // TODO: verify member expr
```
Comment (thk123) [inline unit/analyses/variable-sensitivity/write_stack.cpp]: Done - remove todo

File: unit/analyses/variable-sensitivity/write_stack.cpp [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+          auto stack=write_stackt(in_expr, environment, ns);
+          REQUIRE_FALSE(stack.is_top_value());
+          const exprt &out_expr=stack.to_expression();
+
+          CAPTURE(expr2c(out_expr, ns));
+
+          REQUIRE(out_expr.id()==ID_address_of);
+          const member_exprt &member_exrp=
+            require_exprt::require_member(out_expr.op0(), "comp2");
+          // TODO: verify member expr
```
Comment (thk123) [inline unit/analyses/variable-sensitivity/write_stack.cpp]: Done so remove todo

**Comment by thk123** [comment]
@tautschnig Sorry- I'm not sure I understand your concern. This part of work is essentially a way of taking an AST representing a pointers value and partially evaluating it in the environment so we can represent a pointers value in a symbolic way. This seems to be me an orthogonal problem to branching/non-linear control flow? 

If for example you mean how does it handle:

```C
int a;
int b;
int * c;
if(NONDET())
{
  c = &a;
}
else
{
  c = &b;
}
```

Then this is handled by the `merge` of the `constant_pointer_abstract_objectt` which essentially checks whether they are pointing to the same value, if and only if they are we keep the pointer, otherwise we revert to top. However, this is not modified in this particular PR,

**Comment by tautschnig** [comment]
> This part of work is essentially a way of taking an AST representing a pointers value and partially evaluating it in the environment so we can represent a pointers value in a symbolic way. This seems to be me an orthogonal problem to branching/non-linear control flow?

Yes, I'm starting to understand - sorry, I just didn't quite understand what this PR was touching upon. "A more sophisticated mechanism for storing pointers." was a bit too brief for me to be able to grasp. It's still not clear to me why turning a particular branch of tree (and `exprt`) into a stack makes all the difference, but I'll leave that to others who understand better/have more context.

**Comment by thomasspriggs** [comment]
I am going to close this PR as it appears that work on it has been inactive for an extended period of time. This is not any judgement on how worthwhile this work is. This is part of an effort to reduce the number of open PRs which are not being actively worked on. Getting this PR merged would require someone to be willing to put the work into rebasing and addressing any remaining CI failures, assuming that this work has not been merged as part of another PR. If it is worthwhile to get this merged and someone is willing to put in the work, then the branch should be re-based on the latest version of develop and the PR re-opened.

**Pull request closed without merging**